src/device/random.jl

```diff
@@ -0,0 +1,250 @@
+# Copied from CUDA.jl/src/device/random.jl
+
+
+## random number generation
+
+using Random
+import RandomNumbers
+
+
+# global state
+
+# shared memory with the actual seed, per warp, loaded lazily or overridden by calling `seed!`
+@eval @inline function global_random_keys()
+    ptr = Base.llvmcall(
+        $("""@__zeroinit_global_random_keys = external addrspace($(AS.Local)) global [32 x i32], align 32
+             define i8 addrspace($(AS.Local))* @entry() #0 {
+                 %ptr = getelementptr inbounds [32 x i32], [32 x i32] addrspace($(AS.Local))* @__zeroinit_global_random_keys, i64 0, i64 0
+                 %untyped_ptr = bitcast i32 addrspace($(AS.Local))* %ptr to i8 addrspace($(AS.Local))*
+                 ret i8 addrspace($(AS.Local))* %untyped_ptr
+             }
+             attributes #0 = { alwaysinline }
+          """, "entry"), LLVMPtr{UInt32, AS.Local}, Tuple{})
+    ROCDeviceArray{UInt32,1,AS.Local}((32,), ptr)
+end
+
+# shared memory with per-warp counters, incremented when generating numbers
+@eval @inline function global_random_counters()
+    ptr = Base.llvmcall(
+        $("""@__zeroinit_global_random_counters = external addrspace($(AS.Local)) global [32 x i32], align 32
+             define i8 addrspace($(AS.Local))* @entry() #0 {
+                 %ptr = getelementptr inbounds [32 x i32], [32 x i32] addrspace($(AS.Local))* @__zeroinit_global_random_counters, i64 0, i64 0
+                 %untyped_ptr = bitcast i32 addrspace($(AS.Local))* %ptr to i8 addrspace($(AS.Local))*
+                 ret i8 addrspace($(AS.Local))* %untyped_ptr
+             }
+             attributes #0 = { alwaysinline }
+          """, "entry"), LLVMPtr{UInt32, AS.Local}, Tuple{})
+    ROCDeviceArray{UInt32,1,AS.Local}((32,), ptr)
+end
```

Comment:
Can we use LLVM.jl to build these? E.g. to be robust against IR changes

Reply:
Implemented in the latest commit